src/transformers/models/bark/modeling_bark.py

```diff
@@ -0,0 +1,1372 @@
+        hidden_states = hidden_states + self.mlp(self.ln_2(hidden_states))
+
+        if use_cache:
+            outputs = (hidden_states,) + outputs
+        else:
+            outputs = (hidden_states,) + outputs[1:]
+
+        return outputs  # hidden_states, ((present), attentions)
+
+
+class BarkPreTrainedModel(PreTrainedModel):
+    """
+    An abstract class to handle weights initialization and a simple interface for downloading and loading pretrained
+    models.
+    """
+
+    config_class = BarkConfig
+    base_model_prefix = "bark"
+    supports_gradient_checkpointing = False
+
+    def _init_weights(self, module):
+        if isinstance(module, BarkCausalModule) or isinstance(module, BarkFineAcousticsModule):
+            module.apply(module._init_weights)
+
+    def __init__(self, *inputs, **kwargs):
+        super().__init__(*inputs, **kwargs)
+
+
+class BarkModulePreTrainedModel(PreTrainedModel):
+    """
+    An abstract class to handle weights initialization and a simple interface for downloading and loading pretrained
+    models.
+    """
+
+    config_class = BarkModuleConfig
+    # supports_gradient_checkpointing = True
+    _no_split_modules = ["BarkBlock"]  # TODO: what to do with this?
+
+    def _init_weights(self, module):
+        """Initialize the weights."""
+        if isinstance(module, (nn.Linear,)):
+            # Slightly different from the TF version which uses truncated_normal for initialization
+            # cf https://github.com/pytorch/pytorch/pull/5617
+            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
+            if module.bias is not None:
+                module.bias.data.zero_()
+        elif isinstance(module, nn.Embedding):
+            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
+            if module.padding_idx is not None:
+                module.weight.data[module.padding_idx].zero_()
+        elif isinstance(module, nn.LayerNorm):
+            module.bias.data.zero_()
+            module.weight.data.fill_(1.0)
+
+    def _set_gradient_checkpointing(self, module, value=False):
+        if isinstance(module, BarkCausalModule) or isinstance(module, BarkFineAcousticsModule):
+            module.gradient_checkpointing = value
+
+
+# GPT2-like autoregressive model
+@add_start_docstrings(
+    "Bark sub-module at the core of the semantic and the coarse acoustics sub-models. It is a GPT-2 like autoregressive model with a language modeling head on top.",
+    BARK_MODULE_START_DOCSTRING,
+)
+class BarkCausalModule(BarkModulePreTrainedModel):
+    # TODO: add code sample when checkpoint is added
+    #
+    # @add_code_sample_docstrings(
+    #    checkpoint=_CHECKPOINT_FOR_DOC,
+    #    output_type=BaseModelOutputWithPast,
+    #    config_class=_CONFIG_FOR_DOC,
+    # )
+
+    def __init__(self, config):
+        super().__init__(config)
+        self.config = config
+
+        self._initialize_modules(config)
```

Comment:
We should define the models explicitly here instead of referencing a method, e.g.
```suggestion
self.wte =  nn.Embedding(config.input_vocab_size, config.hidden_size)
self.wpe = nn.Embedding(config.block_size, config.hidden_size)
self.drop = nn.Dropout(float(config.embed_dropout))
self.layers = nn.ModuleList([GPTNeoBlock(config, layer_id=i) for i in range(config.num_layers)])
self.ln_f = nn.LayerNorm(self.embed_dim, eps=config.layer_norm_epsilon)
```

For this class, I would be in favour of defining the 'base' transformer model (no LM head), which we can then re-use for subsequent classes by adding LM heads as required

Reply:
With our following private conversation, is the current (non-modified since our last conversation except for the removal of the _initialize_module_ method) version of the modules work ?